// openrouteservice-js/lib/OrsIsochrones.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _superagent = _interopRequireDefault(require("superagent"));

var _bluebird = _interopRequireDefault(require("bluebird"));

var _OrsUtil = _interopRequireDefault(require("./OrsUtil"));

var _constants = _interopRequireDefault(require("./constants"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _unsupportedIterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _iterableToArray(iter) { if (typeof Symbol !== "undefined" && iter[Symbol.iterator] != null || iter["@@iterator"] != null) return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) return _arrayLikeToArray(arr); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); Object.defineProperty(Constructor, "prototype", { writable: false }); return Constructor; }

var orsUtil = new _OrsUtil["default"]();

var OrsIsochrones = /*#__PURE__*/function () {
  function OrsIsochrones(args) {
    _classCallCheck(this, OrsIsochrones);

    this.meta = null;
    this.args = {};

    if (_constants["default"].propNames.apiKey in args) {
      this.args[_constants["default"].propNames.apiKey] = args[_constants["default"].propNames.apiKey];
    } else {
      // eslint-disable-next-line no-console
      console.log(_constants["default"].missingAPIKeyMsg);
      throw new Error(_constants["default"].missingAPIKeyMsg);
    }

    if (_constants["default"].propNames.host in args) {
      this.args[_constants["default"].propNames.host] = args[_constants["default"].propNames.host];
    }

    if (_constants["default"].propNames.service in args) {
      this.args[_constants["default"].propNames.service] = args[_constants["default"].propNames.service];
    }
  }

  _createClass(OrsIsochrones, [{
    key: "addLocation",
    value: function addLocation(latlon) {
      if (!('locations' in this.args)) {
        this.args.locations = [];
      }

      this.args.locations.push(latlon);
    }
  }, {
    key: "getBody",
    value: function getBody(args) {
      var options = {};

      if (args.restrictions) {
        options.profile_params = {
          restrictions: _objectSpread({}, args.restrictions)
        };
        delete args.restrictions;
      }

      if (args.avoidables) {
        options.avoid_features = _toConsumableArray(args.avoidables);
        delete args.avoidables;
      }

      if (args.avoid_polygons) {
        options.avoid_polygons = _objectSpread({}, args.avoid_polygons);
        delete args.avoid_polygons;
      }

      if (Object.keys(options).length > 0) {
        return _objectSpread(_objectSpread({}, args), {}, {
          options: options
        });
      } else {
        return _objectSpread({}, args);
      }
    }
  }, {
    key: "calculate",
    value: function calculate(reqArgs) {
      // Get custom header and remove from args
      this.customHeaders = [];

      if (reqArgs.customHeaders) {
        this.customHeaders = reqArgs.customHeaders;
        delete reqArgs.customHeaders;
      }

      orsUtil.setRequestDefaults(this.args, reqArgs, true); // eslint-disable-next-line prettier/prettier

      if (!this.args[_constants["default"].propNames.service] && !reqArgs[_constants["default"].propNames.service]) {
        reqArgs.service = 'isochrones';
      }

      orsUtil.copyProperties(reqArgs, this.args);
      var that = this;
      return new _bluebird["default"](function (resolve, reject) {
        var timeout = that.args[_constants["default"].propNames.timeout] || 10000; // eslint-disable-next-line prettier/prettier

        if (that.args[_constants["default"].propNames.apiVersion] === _constants["default"].defaultAPIVersion) {
          // meta should be generated once that subsequent requests work
          if (that.meta == null) {
            that.meta = orsUtil.prepareMeta(that.args);
          }

          that.httpArgs = orsUtil.prepareRequest(that.args);
          var url = orsUtil.prepareUrl(that.meta);
          var postBody = that.getBody(that.httpArgs);
          var authorization = that.meta[_constants["default"].propNames.apiKey];

          var orsRequest = _superagent["default"].post(url).send(postBody).set('Authorization', authorization).timeout(timeout);

          for (var key in that.customHeaders) {
            orsRequest.set(key, that.customHeaders[key]);
          }

          orsRequest.end(function (err, res) {
            if (err || !res.ok) {
              // eslint-disable-next-line no-console
              console.error(err);
              reject(err);
            } else if (res) {
              resolve(res.body || res.text);
            }
          });
        } else {
          // eslint-disable-next-line no-console
          console.error(_constants["default"].useAPIV2Msg);
        }
      });
    }
  }]);

  return OrsIsochrones;
}();

var _default = OrsIsochrones;
exports["default"] = _default;
